renku/core/management/datasets.py

```diff
@@ -247,6 +250,51 @@ def with_dataset(self, name=None, create=False):
 
         dataset.to_yaml()
 
+    @contextmanager
+    def with_dataset_provenance(self, name=None, create=False):
+        """Yield a dataset's metadata from dataset provenance."""
+        dataset = self.load_dataset_from_provenance(name=name)
+        clean_up_required = False
+        dataset_ref = None
+        path = None
+
+        if dataset is None:
+            if not create:
+                raise errors.DatasetNotFound(name=name)
+
+            clean_up_required = True
+            dataset, path, dataset_ref = self.create_dataset(name=name)
+        elif create:
+            raise errors.DatasetExistsError('Dataset exists: "{}".'.format(name))
+        else:
+            dataset = dataset.to_dataset(self)
+
+        dataset_path = self.path / self.data_dir / dataset.name
+        dataset_path.mkdir(parents=True, exist_ok=True)
+
+        try:
+            yield dataset
+        except Exception:
```

Comment:
Should we use `RenkuException` here?

Reply:
No, this should catch non-Renku exceptions to do a cleanup.